pkg/controllers/webhookcerts/cert_requester.go

```diff
@@ -0,0 +1,412 @@
+		WithFilteredEventsInformers(nameFilter(csrName), csrInformers.Informer()).
+		WithFilteredEventsInformers(nameFilter(certSecretName), secretInformers.Informer()).
+		ToController("WebhookAuthenticatorCertRequester", eventsRecorder.WithComponentSuffix("webhook-authenticator-cert-requester"))
+}
+
+func (c *webhookAuthenticatorCertRequester) syncWithConditions(ctx context.Context, syncCtx factory.SyncContext) error {
+	progressing, err := c.sync(ctx, syncCtx)
+	if err != nil {
+		return err
+	}
+
+	// nil == keep the current progressing status
+	if progressing == nil {
+		return nil
+	}
+
+	var progressingText string
+	progressingStatus := operatorv1.ConditionFalse
+	progressingReason := "CertificateAvailable"
+	if *progressing {
+		progressingText = "waiting for CSR to be signed"
+		progressingStatus = operatorv1.ConditionTrue
+		progressingReason = "WaitingForCertificate"
+	}
+
+	_, _, statusErr := v1helpers.UpdateStatus(c.operatorClient,
+		v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:    "AuthenticatorCSRProgressing",
+			Status:  progressingStatus,
+			Reason:  progressingReason,
+			Message: progressingText,
+		}))
+	return statusErr
+}
+
+func (c *webhookAuthenticatorCertRequester) sync(ctx context.Context, syncCtx factory.SyncContext) (progressing *bool, err error) {
+	key, secretComplete, shouldResubmit, err := c.checkKeyCertSecret(ctx, syncCtx.Recorder())
+	if err != nil {
+		return nil, err
+	}
+
+	if secretComplete {
+		return pbool(false), err
+	}
+
+	if !shouldResubmit {
+		var cert []byte
+		cert, shouldResubmit, err = c.handleCSR(key, syncCtx.Recorder())
+		if err != nil {
+			return nil, err
+		}
+
+		if cert != nil {
+			c.attachCertToSecret(cert, syncCtx.Recorder())
+			// we've got the cert, stop progressing
+			return pbool(false), nil
+		}
+
+		if !shouldResubmit { // keep on waiting for the current CSR
+			return pbool(true), nil
+		}
+
+	}
+
+	// either we did not have a key or the CSR failed
+	if shouldResubmit {
+		if err := c.resubmitCSR(ctx, key); err != nil {
+			return nil, err
+		}
+		return pbool(true), nil
+	}
+
+	return pbool(false), nil
+}
+
+func (c *webhookAuthenticatorCertRequester) checkKeyCertSecret(ctx context.Context, recorder events.Recorder) (key *rsa.PrivateKey, secretComplete, shouldResubmit bool, err error) {
+	key, cert, err := c.getKeyCertFromSecret()
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, false, false, err
+	}
+
+	if key != nil && cert != nil {
+		// check that the certificate is not expiring
+		certValidityDuration := cert.NotAfter.Sub(cert.NotBefore)
+		if time.Now().Before(cert.NotBefore.Add(time.Duration(0.8 * float64(certValidityDuration)))) {
+			return key, true, false, nil
+		}
+		// set key == nil to trigger new key generation
+		key = nil
+
+		expiryMsg := fmt.Sprintf("the authenticator's certificate is almost expired (%s), will schedule a new CSR creation", cert.NotAfter.String())
+		recorder.Eventf("ClientCertExpiringSoon", expiryMsg)
```

Comment:
this line has potential to keep creating events continuously. Only create events on an edge, not based on state like here. What is the edge?

Reply:
Would move it after a successful createSecretWithNewKey.

Reply:
The certificate gets removed in the next step.

I arbitrarily picked 80% of the lifetime of the certificate to be the edge. I don't think there's a chance we get here at the very moment when the the certificate reaches that lifetime edge.

Reply:
I fixed this by adding this event call in defer only if `err == nil`